src/graphics/canvas-layer.js

```diff
@@ -193,7 +222,7 @@ Crafty.canvasLayerObject = {
             for (j = 0, len = q.length; j < len; ++j) {
                 obj = q[j];
 
-                if (dupes[obj[0]] || !obj._visible || !obj.__c.Canvas)
+                if (dupes[obj[0]] || !obj._visible || (obj._drawLayer !== this) )
```

Comment:
One hashmap per layer and one dedicated hashmap for collision stuff would be the dream (I stumbled upon this "inconvenience" when separating isometric entities into their display part and their game-logic part).   
The only downside I see to this, is that multiple entries (possibly only 2) would have to be managed per entity (so a small performance impact). However, the performance gain (by not considering irrelevant entries inside a layer) could easily outweigh that.